arch/lkl/kernel/setup.c

```diff
@@ -163,14 +163,22 @@ long lkl_sys_halt(void)
 
 	is_running = false;
 
-	lkl_cpu_wait_shutdown();
+	/**
+	 * TODO: We may have cloned host threads that did not terminate,
+	 * and prevent us from shutting down the LKL CPU.
+	 *
+	 * As a workaround, we do not shutdown all threads.
```

Comment:
I think the correct thing to do here is run the scheduler to wake up all threads and allow them to exit the kernel but modify the `switch_to` routine so that during shutdown it doesn't do sem_down on cloned threads.

Reply:
When a process does a `do_exit()`, LKL normally kills all of the userspace threads (both deallocating them in lthreads and LKL) but not always. In rare cases, e.g. after a few 100s of JVM runs, we have one cloned thread hanging around, which motivates the above workaround.